compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/declarations/IrValueParameter.kt

```diff
@@ -38,6 +38,11 @@ abstract class IrValueParameter : IrValueDeclaration(), IrSymbolDeclaration<IrVa
     // Once we are able to load any top-level declaration from klib this hack should be deprecated and removed.
     abstract val isHidden: Boolean
 
+    // Parameters are not assignable by default. However, in the IR, assignments to parameters are used
+    // to implement default argument stubs. If the default value is to be used, the value is assigned
+    // to the parameter.
+    abstract val isAssignable: Boolean
```

Comment:
I would suggest hoist this property into `IrValueDeclaration`

Reply:
We decided to not do that. The reason is that we cannot currently check that we only perform assignments to variables marked as `isVar`. Therefore, we can use `isVar` but then there is no benefit as nothing will be checked. For `IrVariable`, `isVar` is used for other things during codegen; but there is no checking that only `isVar` variables are assigned.

So, we ended up here, so that we could introduce the check that parameters are only assigned if they are explicitly marked as assignable.

Reply:
Come to think of it we already do check via 'is' checks on the IrValueDeclaration, so we could use `isVar` and hoist it to `IrValueDeclaration` and then only check that a value is `isVar` on assignment if it is a value parameter. The only remaining issue that I have is that value parameters cannot be `var` and therefore I like the `isAssignable` bit on value parameters better. But we *could* use `isVar` if that is preferable. 

Thoughts: @udalov @romanart ?

Reply:
I have a couple of thinks here:
 - I would like that we have only one flag which says whether value is assignable or mutable.
 - It should be checked likely without explicit cast in `IrSetValue*` 
 - Probably `isVar` isn't the best name for that flag.

Reply:
Unfortunately, we cannot have all three of those. We can have one flag called `isAssignable`, but then we need the type checks and casts in `IrSetValue`. In the JVM_IR backend, we do assign to variables that have `isVar=false` and we have to in order to generate code that can be inlined by the bytecode inliner I believe. So, if we unify on one flag, we need the type checking and casts. Currently, there are no checks that we only use IrSetVariable with variables that have `isVar=true` and the reason is that we actually do have to do that. :-\

Reply:
As discussed offline, I have moved `isAssignable` to `IrValueDeclaration` and have it always be true for `IrVariable` so that we can unconditionally assert in `IrSetValueImpl` that the declaration is assignable.